lib/codec.js

```diff
@@ -0,0 +1,165 @@
+  // calculate the final values for the term codec map
+  const terms = Array.from(termCodecMap.keys()).sort();
+  terms.forEach((term, index) => {
+    if(encode) {
+      termCodecMap.get(term).value = index;
+    } else if(decode) {
+      termCodecMap.get(term).value = term;
+      termCodecMap.set(index, termCodecMap.get(term));
+    }
+  });
+
+  // overlay the app term map on the existing term encoding map
+  if(decode && appTermMap.size > 0) {
+    const codecMap = new CodecMapCodec();
+    codecMap.set({value: appTermMap, termCodecMap});
+    appTermMap = codecMap.decodeCBOR();
+  }
+
+  // clear terms used during encoding
+  if(decode) {
+    terms.forEach(term => termCodecMap.delete(term));
+  }
+
+  for(const [key, value] of appTermMap.entries()) {
+    const codec = (encode) ? termCodecs.get(value) : value;
+    if(codec === undefined) {
+      throw new CborldError(
+        'ERR_UNKNOWN_TERM_CODEC',
+        `Unknown codec '${value}' specified for term '${key}'.`);
+    }
+    termCodecMap.get(key).codec = codec;
+  }
+
+  return termCodecMap;
+}
+
+// Generates a compression map given a JSON-LD Context
+function _generateTermEncodingMap(context) {
+  const termCodecMap = new Map();
+
+  // sanity check that context is a valid JSON-LD context object
+  if(!context['@context']) {
+    throw new CborldError(
+      'ERR_INVALID_JSONLD_CONTEXT',
+      'The JSON-LD Context does not contain a top-level @context value: ' +
+      JSON.stringify(context));
+  }
+
+  // FIXME: Ensure that all JSON-LD keywords that could be keys are accounted
+  // for in the map -- e.g., @id is currently missing
+  termCodecMap.set('@codec', {
+    codec: CodecMapCodec
+  });
+  termCodecMap.set('@context', {
+    codec: ContextCodec
+  });
+  termCodecMap.set('@type', {
+    codec: UrlCodec
+  });
+
+  for(const [key, value] of Object.entries(context['@context'])) {
+    termCodecMap.set(key, {
+      codec: SimpleTypeCodec
+    });
+
+    if(typeof value === 'string') {
+      termCodecMap.get(key).codec =
+        (value === '@type') ? UrlCodec : SimpleTypeCodec;
+    } else if(typeof value === 'object') {
+      // FIXME: Add support for global and local type-specific compressors
+      let codec = SimpleTypeCodec;
+      switch(value['@type']) {
+        case '@id':
+        case '@vocab':
+          codec = UrlCodec;
+          break;
+        // FIXME: Keep track of prefixes and see if xsd matches
+        // http://www.w3.org/2001/XMLSchema#
+        case 'xsd:dateTime':
+          codec = XsdDateTimeCodec;
+          break;
+          // FIXME: Keep track of prefixes and see if sec matches
+          // https://w3id.org/security#
+        case 'sec:base64Pad':
+          codec = Base64PadCodec;
+      }
+      termCodecMap.get(key).codec = codec;
+
+      if(value['@context']) {
+        const scopedTermEncodingMap = _generateTermEncodingMap(value);
+        scopedTermEncodingMap.forEach(
+          (value, key) => termCodecMap.set(key, value));
```

Comment:
Determining the appropriate compressor for these situations is going to be challenging -- not sure if this general algorithm works. Might need to hook into jsonld.js expansion to do compression properly.

Reply:
Raised issue https://github.com/digitalbazaar/cborld/issues/2 to track this bug.